### DISCUSSION TITLE
Are plugin and charon incompatible between different strongSwan versions?

### DISCUSSION THREAD

**Discussion by wataash** [strongswan/strongswan#1435] [Q&A] [answered]
I built strongSwan **5.9.8** and copied the save-keys plugin to the system (Ubuntu 20.04), where strongSwan **5.8.2** was installed. Then charon was unable to startup and died with SEGV.

```sh
/usr/lib/ipsec/charon --version
# Linux strongSwan 5.8.2

sudo cp strongswan/build-5.9.8/src/libcharon/plugins/save_keys/.libs/libstrongswan-save-keys.so /usr/lib/ipsec/plugins/libstrongswan-save-keys.so
/usr/lib/ipsec/charon
# C 21:43:35 00[DMN] Starting IKE charon daemon (strongSwan 5.8.2, Linux 5.15.39+, x86_64)
# fish: Job 2, '/usr/lib/ipsec/charon' terminated by signal SIGSEGV (Address boundary error)
```

When I built strongSwan 5.8.2, the same version as the system, the plugin works fine.

```sh
sudo cp strongswan/build-5.8.2/src/libcharon/plugins/save_keys/.libs/libstrongswan-save-keys.so /usr/lib/ipsec/plugins/libstrongswan-save-keys.so
/usr/lib/ipsec/charon
# C 12:15:34 00[DMN] Starting IKE charon daemon (strongSwan 5.8.2, Linux 5.15.39+, x86_64)
# C 12:15:34 00[DMN] !!
# C 12:15:34 00[DMN] !! WARNING: SAVING IKE AND ESP KEYS TO '/home/wsh/.config/wireshark/'
# C 12:15:34 00[DMN] !!
# C 12:15:34 00[LIB] plugin 'save-keys': loaded successfully
# C 12:15:34 00[LIB] plugin 'aesni': loaded successfully
# ...
```

I found out that 5.9.5 broke the ABI compatibility of `library_t` [when `library_t::metadata` was added](https://github.com/strongswan/strongswan/commit/531335ad20e8aab902e5b3aa422861b2e910ee9d#diff-ee26ce01effe2138c7e177e7930debc18efd358639196cb5e6889ea0bf82463fR196). In this case, `lib->settings` became `NULL`:

```c
#0  0x00007f37fc98bd8d in save_keys_listener_create () at /home/wsh/qc/strongswan/build/../src/libcharon/plugins/save_keys/save_keys_listener.c:402
	// .path = lib->settings->get_str(lib->settings, // lib->settings: NULL
#1  0x00007f37fc98b3bd in save_keys_plugin_create () at /home/wsh/qc/strongswan/build/../src/libcharon/plugins/save_keys/save_keys_plugin.c:95
#2  0x00007f37fca66159 in ?? () from target:/usr/lib/ipsec/libstrongswan.so.0
#3  0x00007f37fca666df in ?? () from target:/usr/lib/ipsec/libstrongswan.so.0
#4  0x00007f37fc9a5d45 in ?? () from target:/usr/lib/ipsec/libcharon.so.0
#5  0x000055ad33c58c51 in ?? ()
#6  0x00007f37fc766083 in __libc_start_main (main=0x55ad33c584e0, argc=1, argv=0x7ffc6ee2ee98, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffc6ee2ee88) at ../csu/libc-start.c:308
#7  0x000055ad33c5928e in ?? ()
```

So, the plugins are compatible with the charon daemon only if both the plugin and charon are built in the same version, right? (it doesn't seem to be documented)

**Comment by strongX509** [comment, marked as answer]
Yes, you are right, the libstrongswan ABI is not guaranteed to be stable and can change from time to time. It is recommended to always use the same strongSwan version for the executables, libraries and the plugins. The *`--enable-integrity-test`* configure option as documented in 

https://docs.strongswan.org/docs/5.9/features/integrityTests.html

guarantees to automatically detect old, incompatible or corrupted strongSwan components.

**Comment by wataash** [comment]
Thanks!
